File: 2ndSemester/AlgorithimsAndDataStructuresII/Practice/PraticalAssignment01/Q02-PalindromoC/palindromo.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

//funcao que resolve o problema
bool checkPalindromo(char entrada[]){
  bool resultado = true;
  int tamanho = strlen(entrada);
  int i = 0;
  do{
    if(entrada[i] != entrada[tamanho - 1 -i]){
      resultado = false;
    }
    i++;
  }while(i < tamanho/2 && resultado != false);
  return resultado;
}



//funcao que verifica o fim do loop 
bool checkFim(char s[]){
  return (strlen(s) >= 3 && s[0] == 'F' && s[1] == 'I' && s[2] == 'M');
}

//funcao principal
int main(void){
  char entrada[1000];
  bool fimPrograma;
  do{
    scanf(" %[^\n]", entrada);
    fimPrograma = checkFim(entrada);
    if(!fimPrograma){
      if(checkPalindromo(entrada)){
        printf("SIM\n");
      }
      else {
        printf("NAO\n");
      }
    }
    

  }while(!fimPrograma);
  return 0;
}

File: 2ndSemester/AlgorithimsAndDataStructuresII/Practice/PraticalAssignment01/Q11-PalindromoRecursivoJava/Palindromo.java
class Palindromo {
  public static boolean checkPalindromoRecursivo(String entrada, int i){
    boolean resultado = true;
    int tamanho = entrada.length();
    
    if(entrada.charAt(i) != entrada.charAt(tamanho - 1- i))
      resultado = false;
    else if(i < tamanho/2)
      resultado = checkPalindromoRecursivo(entrada, i + 1);

    return resultado;
  }
  //metodo base para a recursao
  public static boolean checkPalindromo(String entrada){
    return checkPalindromoRecursivo(entrada, 0);
  }

  public static boolean checkFim(String entrada){
    return (entrada.length() == 3 && entrada.charAt(0) == 'F' && entrada.charAt(1) == 'I' && entrada.charAt(2) == 'M');
  }
  //metodo principal
  public static void main(String[] args){
    String entrada;
    do{
      entrada = MyIO.readLine();
      if(checkFim(entrada) == false){
        if(checkPalindromo(entrada)){
          MyIO.println("SIM");
        }
        else{
          MyIO.println("NAO");
        }
      }
    }while(checkFim(entrada) == false);
  } 
}
